# Task Distribution
| Member | ID | Tasks |
|---|---|---|
|Bhavika| - | - | 
|Alora| - | - |
|Andrea| - | - |

# Import Libraries

In [2]:
%pip install tweepy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [3]:
import tweepy
import sys
import csv
import pandas as pd
import re

# Scraping Tweets

In [4]:
# Keys for authentication
consumer_key = "ONLQ0vLSAXM2tZDnckominEcK"
consumer_secret = "MpP7XurdGn1wpufX3rrfQgrs4AROnaw9ZiiZzKN6exr3zEDlN6"
access_token = "1274112117738737664-J2DjjoqqzN11CNp8bpnYzVp1dUYKM4"
access_token_secret = "3coYFJVCqHnOmD2m9Qt6zs7j1I24lhlPTU71eUOq4Zqg9"
bearer_token = "AAAAAAAAAAAAAAAAAAAAALJslQEAAAAA7EWdkTwrVDM44LIuewNascHjvoY%3DeXmhQyuXgo8zXJiSTMfqtmPA393HHO2W7nSrk0bK6b67SuJeb0"

In [5]:
# Authenticate
auth = tweepy.OAuth1UserHandler(
    consumer_key, consumer_secret, access_token, access_token_secret)

api = tweepy.API(auth)

In [6]:
def scrape_tweets(query: str, amount: int) -> list:
    """Scrapes a specified number of tweets based on given query and location.

    Method from https://www.sahilfruitwala.com/guide-to-extract-tweets-using-tweepy#how-to-retrieve-specific-number-of-tweets-using-tweepy 
    
    Args:
        query (str): The query
    Returns:
        list (str): List of contents of tweets.
    """
    extracted_tweets = []
    for tweet in tweepy.Cursor(api.search_tweets, query, count=100, tweet_mode="extended", result_type="recent").items(amount):
        extracted_tweets.append(tweet.full_text)
    return extracted_tweets

In [7]:
def write_to_csv(tweets: list, csv_name: str) -> None:
    """Appends the list of tweets to csv.
    
    Largely copied from https://gist.github.com/anku255/0cebd75cce675f2b56de1ef48ec06575.

    """
    print("Writing {0} tweets to {1}".format(len(tweets), csv_name))
    tweets_for_csv = [[tweet.encode("utf-8")] for tweet in tweets]
    with open(csv_name, "a+") as file:
        writer = csv.writer(file, delimiter=",")
        writer.writerows(tweets_for_csv)


The queries for scraping tweets.

In [8]:

queries = [
    ["(metaverse OR meta verse OR #metaverse OR #meta #verse) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(innovation OR #innovation OR innovate OR innovative OR #innovate OR #innovative) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(sustainability OR #sustainability OR sustainable) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technology footprint OR technology OR #technologyfootprint) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(artificial intelligence OR ai OR #ai OR #artificialintelligence) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(ethics OR #ethics OR ethical OR #ethical) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(inflation OR #inflation) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cop 28 OR cop28 OR #cop28 OR #cop #28) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(museum of the future OR @museumofthefuture OR #museumofthefuture OR #museum #of #the #future) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(bit coin OR bitcoin OR #bitcoin OR #crypto OR crpyto OR cryptocurrency) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cyber attack OR #cyberattack OR #cyber OR #cyberattacks OR #cyber #attack) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(robot OR robots OR #robots OR #robot) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(drone OR drones OR #drone OR #drones) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(hacker OR hacking OR #hacker OR #hacking OR #hack) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(chatgpt OR #chatgpt OR #chat #gpt OR chat gpt) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cybersecurity OR #cybersecurityOR #cybersec OR cybersec OR cyber security OR #cyber #security) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(sustainabletech OR sustainable ai OR sustainable technology OR #sustainabletech OR #sustainableai OR #sustainabletechnology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(computing OR quantum computing OR #quantum #computing OR #computing) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(automation OR #automation) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(distributed cloud OR #distributedcloud OR #distributed #cloud) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(big data OR #big #data OR bigdata OR #bigdata) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(ar OR #ar OR #augmentedreality OR #augmented #reality) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(data mining OR #data #mining OR #datamining) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(software OR #software) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(tech OR technology OR #tech OR #technology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(digital OR digital transformation OR #digital #transformation OR #digitaltransformation) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(blockchain OR #blockchain) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(coders_hq OR coders hq OR #coders OR #coders #hq OR @coders_hq) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(nft OR non-fungible token OR non fungible token OR nfts OR #nfts OR #nonfungibletoken OR #nft) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(binance OR @binance OR #binance) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(uae hackathon OR hackathon OR #hackathon) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(internet of things OR IoT OR #iot OR #internet #of #things OR #internetofthings) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(software engineering OR #softwareengineering) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(#coding OR coding) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(@uaeai OR #uaeai Or #uae #ai) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technology OR tech OT #tech Or #technology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(nft OR #nft OR #nfts OR nfts) AND (place_country:AE OR uae) lang:en -filter:retweets"]   ,
    ["(UAE codes OR #UAE_codes OR #UAEcodes) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(Microsoft Hololens OR Hololens OR #Hololens) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(GITEX or #gitex) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(information OR #information) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(internet OR #internet) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(computer OR science OR computer science OR #computerscience OR #computer OR #science) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technological OR #technological) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(engineering OR #engineering) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(computing OR #computing) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(systems OR #systems) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(electronics OR #electronics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(application OR #application OR app OR #app) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(robotics OR #robotics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(high tech OR high technology OR #hightech OR #hightechnology OR #high #technology OR #high #tech) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(nanotech OR nano tech OR nanotechnology OR nano technology OR #nanotech OR #nano #tech OR #nano #technology OR #nanotechnology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(biotech OR bio tech OR biotechnology OR bio technology OR #biotech OR #bio #tech OR #bio #technology OR #biotechnology) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(code OR #code) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(automation OR #automation OR automate OR #automate) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(virtual OR #virtual OR online OR #online) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(technological advancements OR #technological #advancements OR #advancements OR #techadvancements) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(bionics OR #bionics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cyberscience OR #cyberscience) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(google OR #google OR @google) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(oracle OR #oracle OR @oracle) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(ibm OR #ibm OR @ibm) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(bionics OR #bionics) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(sap OR #sap) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(global knowledge OR #globalknowledge OR #global #knowledge) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(amazon OR #amazon OR @amazon) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(hewlett-packard OR Hewlett Packard OR HP OR #hewlett-packard OR #Hewlett #Packard OR #HP) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cisco OR #cisco OR @cisco) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(dell OR #dell OR @dell) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(cisco OR #cisco OR @cisco) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(@WorldGovSummit OR #WGS OR #WGS2023 OR World Goverment Summit) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(masdar OR #masdar OR @masdar) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(renewable energy OR clean energy OR #renewableenergy OR #cleanenergy) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(machine learning OR #machinelearning OR ML OR #ML) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(recommender systems OR #RS OR #recommender #systems OR recommender engine) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(deep learning OR #deeplearning) AND (place_country:AE OR uae) lang:en -filter:retweets"],
    ["(research OR #research) AND (place_country:AE OR uae) lang:en -filter:retweets"],
]

Commented out below to prevent re-running code.

In [9]:
# """Scraping tweets and writing them to dataset.csv"""
# csv_name="dataset.csv"
# total = 0
# for query in queries:
#     tweets = scrape_tweets(query=query, amount=500)
#     total += len(tweets)
#     print("Found {0} tweets related to the query {1}".format(len(tweets), query))
#     write_to_csv(tweets=tweets,csv_name=csv_name)
# print("Done. {0} rows.".format(total))

# Data Cleaning

This section cleans the scraped tweets dataset of duplicate tweets, URLs, @'s and #s.

Note: 
- Do not modify original csv `dataset.csv`. The dataset with no duplicates is stored in `updated_dataset.csv`

In [10]:
# Dropping duplicates
dataset = pd.read_csv("dataset.csv", header=None)
dataset.rename(columns={0: 'Text'}, inplace=True)


In [11]:
dataset

Text
0      b'Joined UAE Minister &amp; President-Designat...
1      b'Foreign Secretary @AmbVMKwatra participated ...
2      b'As the UAE marks National Environment Day, w...
3      b'RT @KhalifaAlgaz: Did you know that the UAE ...
4      b'RT @tawhidChtioui: \xf0\x9f\x9b\xb0\xef\xb8\...
...                                                  ...
21974  b'Send our special gifts to your loved ones!\x...
21975  b'Trainer code: 2735 9457 4031\nfrom \xf0\x9f\...
21976  b"We are looking for ICY SNOW , SANDSTORM .\nI...
21977  b'Make #MONEY from #home through PC #APP \nMak...
21978  b'#Earn Money from #home through PC #APP\nwith...

[21979 rows x 1 columns]

In [12]:
updated_dataset = dataset.drop_duplicates().copy()
updated_dataset


Text
0      b'Joined UAE Minister &amp; President-Designat...
1      b'Foreign Secretary @AmbVMKwatra participated ...
2      b'As the UAE marks National Environment Day, w...
3      b'RT @KhalifaAlgaz: Did you know that the UAE ...
4      b'RT @tawhidChtioui: \xf0\x9f\x9b\xb0\xef\xb8\...
...                                                  ...
21888  b'@m_ut67 @modgovae @Forsan_UAE Dear sir pleas...
21889  b'@UAE_BARQ Dear sir please my help you sir my...
21890  b'@UAE_BARQ Dear sir please my help you sir my...
21915  b"#PhonePe launched a service on Tuesday that ...
21962  b'@WeLoveIndia7 Saudi k log bhikhari h ? \nIn ...

[5878 rows x 1 columns]

## Removing URLs, @'s and hashtags

In [13]:
tweets = updated_dataset['Text'].copy()
updated_dataset['Cleaned Text'] = tweets.str.replace(r'@[^\s]+|#[^\s]+|https?:\/\/\S+|www\.\S+', '',regex=True)



## Removing Duplicates again

In [14]:
updated_dataset = updated_dataset.drop('Text', axis=1)
updated_dataset = updated_dataset.drop_duplicates()
updated_dataset.shape

(5398, 1)

In [15]:
updated_dataset.to_csv('updated-dataset.csv')

In [16]:
updated_dataset = pd.read_csv('updated-dataset.csv', index_col=0)
updated_dataset

Cleaned Text
0      b'Joined UAE Minister &amp; President-Designat...
1      b'Foreign Secretary  participated in a meeting...
2      b'As the UAE marks National Environment Day, w...
3      b'RT  Did you know that the UAE is developing ...
4      b'RT  \xf0\x9f\x9b\xb0\xef\xb8\x8f UAE lunar r...
...                                                  ...
21888  b'   Dear sir please my help you sir my two ki...
21889  b' Dear sir please my help you sir my two kidn...
21890  b' Dear sir please my help you sir my two kidn...
21915  b" launched a service on Tuesday that will all...
21962  b' Saudi k log bhikhari h ? \nIn Middle East, ...

[5398 rows x 1 columns]

# Data Labelling


## TextBlob

For data labelling, we decided to use first use TextBlob. TextBlob calculates the subjectivity and polarity of a text to classify the text as 'Positive', 'Negative' or 'Neutral'

In [17]:
%pip install TextBlob


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [18]:
from textblob import TextBlob

In [19]:
#Function to get subjectivity
def getSubjectivity(text):
   return TextBlob(text).sentiment.subjectivity
  
 
#Function to get the polarity
def getPolarity(text):
    return TextBlob(text).sentiment.polarity
   
#Create two new columns ‘Subjectivity’ & ‘Polarity’
updated_dataset['TextBlob Subjectivity'] =    updated_dataset['Cleaned Text'].apply(getSubjectivity)
updated_dataset['TextBlob Polarity'] = updated_dataset['Cleaned Text'].apply(getPolarity)
def getAnalysis(score):
  if score < 0:
    return 'Negative'
  elif score == 0:
    return 'Neutral'
  else:
    return 'Positive'

updated_dataset['TextBlob Sentiment'] =  updated_dataset['TextBlob Polarity'].apply(getAnalysis)


In [20]:
updated_dataset['TextBlob Sentiment'].value_counts()

Positive    2643
Neutral     2276
Negative     479
Name: TextBlob Sentiment, dtype: int64

It can be observed that TextBlob classified the majority of tweets to be Positive. To compare, we also decided to use Vader as well to observe if there was a major difference in classification.

## Vader

In [21]:
%pip install vaderSentiment

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [22]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

sentiment = SentimentIntensityAnalyzer()
def getSentiment(text):
    return sentiment.polarity_scores(text)
    
updated_dataset['Vader Analysis'] = updated_dataset['Cleaned Text'].apply(getSentiment)

def sentimentAnalysis(sentiment_dict):
    if sentiment_dict['compound'] >= 0.05:
        return 'Positive'
    elif sentiment_dict['compound'] <= -0.05:
        return 'Negative'
    else:
        return 'Neutral'

updated_dataset['Vader Sentiment'] = updated_dataset['Vader Analysis'].apply(sentimentAnalysis)

In [23]:
updated_dataset[['Vader Analysis', 'Vader Sentiment']]


Vader Analysis Vader Sentiment
0      {'neg': 0.0, 'neu': 0.638, 'pos': 0.362, 'comp...        Positive
1      {'neg': 0.0, 'neu': 0.816, 'pos': 0.184, 'comp...        Positive
2      {'neg': 0.0, 'neu': 0.766, 'pos': 0.234, 'comp...        Positive
3      {'neg': 0.0, 'neu': 0.874, 'pos': 0.126, 'comp...        Positive
4      {'neg': 0.0, 'neu': 0.819, 'pos': 0.181, 'comp...        Positive
...                                                  ...             ...
21888  {'neg': 0.073, 'neu': 0.726, 'pos': 0.202, 'co...        Positive
21889  {'neg': 0.073, 'neu': 0.726, 'pos': 0.202, 'co...        Positive
21890  {'neg': 0.074, 'neu': 0.719, 'pos': 0.206, 'co...        Positive
21915  {'neg': 0.043, 'neu': 0.774, 'pos': 0.183, 'co...        Positive
21962  {'neg': 0.196, 'neu': 0.777, 'pos': 0.028, 'co...        Negative

[5398 rows x 2 columns]

In [24]:
updated_dataset['Vader Sentiment'].value_counts()

Positive    3335
Neutral     1691
Negative     372
Name: Vader Sentiment, dtype: int64

# Data Preprocessing

**changed approach → first data cleaning + tokenization then stemming OR lemmatization. (comparison)**

## Additional Data Cleaning

### Removing Non-English Characters

In [25]:
def nonenglish(tweet):
    tweet = re.sub(r'\\x[a-zA-Z0-9]+', '', tweet) # tweets in different language changed to hex code \xhh
    tweet = re.sub(r'\\n', '', tweet) # removing new line character as well
    return tweet

updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: nonenglish(tweet).lower())
updated_dataset['Cleaned Text'].head()


0    b'joined uae minister &amp; president-designat...
1    b'foreign secretary  participated in a meeting...
2    b'as the uae marks national environment day, w...
3    b'rt  did you know that the uae is developing ...
4    b'rt   uae lunar rover will test 1st artificia...
Name: Cleaned Text, dtype: object

### Removing Stopwords

In [26]:
from sklearn.feature_extraction import text

# contains 318 stopwords, including but not limited to articles and prepositions
stopwords = text.ENGLISH_STOP_WORDS

updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: ' '.join([text for text in tweet.split(' ') if text not in stopwords]))
updated_dataset['Cleaned Text'].head()

0    b'joined uae minister &amp; president-designat...
1    b'foreign secretary  participated meeting foca...
2    b'as uae marks national environment day, remai...
3    b'rt  did know uae developing arabic chatgpt u...
4    b'rt   uae lunar rover test 1st artificial int...
Name: Cleaned Text, dtype: object

### Removing Symbols and Punctuations

In [27]:
import string

punctuations = string.punctuation
print(punctuations)

updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: tweet.translate({ord(punc): None for punc in punctuations}))
updated_dataset['Cleaned Text'].head()

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


0    bjoined uae minister amp presidentdesignate  d...
1    bforeign secretary  participated meeting focal...
2    bas uae marks national environment day remain ...
3    brt  did know uae developing arabic chatgpt us...
4    brt   uae lunar rover test 1st artificial inte...
Name: Cleaned Text, dtype: object

### Removing Numbers

In [28]:
updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: re.sub('[0-9]+', '', tweet))
updated_dataset['Cleaned Text'].head()

0    bjoined uae minister amp presidentdesignate  d...
1    bforeign secretary  participated meeting focal...
2    bas uae marks national environment day remain ...
3    brt  did know uae developing arabic chatgpt us...
4    brt   uae lunar rover test st artificial intel...
Name: Cleaned Text, dtype: object

### Removing 'b' and 'RT'

In [29]:
updated_dataset['Cleaned Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: re.sub('bRT|b', '', tweet))
updated_dataset['Cleaned Text'].head()

0    joined uae minister amp presidentdesignate  dr...
1    foreign secretary  participated meeting focal ...
2    as uae marks national environment day remain c...
3    rt  did know uae developing araic chatgpt usin...
4    rt   uae lunar rover test st artificial intell...
Name: Cleaned Text, dtype: object

## Tokenization

In [30]:
# as data has been completely cleaned, we can simply tokenize by converting strings to lists
updated_dataset['Tokenized Text'] = updated_dataset['Cleaned Text'].apply(lambda tweet: tweet.split())
updated_dataset['Tokenized Text'].head()

0    [joined, uae, minister, amp, presidentdesignat...
1    [foreign, secretary, participated, meeting, fo...
2    [as, uae, marks, national, environment, day, r...
3    [rt, did, know, uae, developing, araic, chatgp...
4    [rt, uae, lunar, rover, test, st, artificial, ...
Name: Tokenized Text, dtype: object

## Stemming

In [31]:
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /home/andrea/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [32]:
stemming = nltk.stem.PorterStemmer()

updated_dataset['Stemmed Text'] = updated_dataset['Tokenized Text'].apply(lambda tweet: [stemming.stem(text) for text in tweet])
updated_dataset['Stemmed Text'].head()

0    [join, uae, minist, amp, presidentdesign, dr, ...
1    [foreign, secretari, particip, meet, focal, po...
2    [as, uae, mark, nation, environ, day, remain, ...
3    [rt, did, know, uae, develop, araic, chatgpt, ...
4    [rt, uae, lunar, rover, test, st, artifici, in...
Name: Stemmed Text, dtype: object

## Lemmatization

In [33]:
lemmatizing = nltk.stem.WordNetLemmatizer()

updated_dataset['Lemmatized Text'] = updated_dataset['Tokenized Text'].apply(lambda tweet: [lemmatizing.lemmatize(text) for text in tweet])
updated_dataset['Lemmatized Text'].head()

0    [joined, uae, minister, amp, presidentdesignat...
1    [foreign, secretary, participated, meeting, fo...
2    [a, uae, mark, national, environment, day, rem...
3    [rt, did, know, uae, developing, araic, chatgp...
4    [rt, uae, lunar, rover, test, st, artificial, ...
Name: Lemmatized Text, dtype: object

In [34]:
# Encode TextBlob Sentiments into numbers
updated_dataset['Encoded_TextBlob_Sentiment'] = updated_dataset['TextBlob Sentiment'].apply(lambda x: 1 if x=='Positive' else -1 if x=='Negative' else 0)

In [35]:
updated_dataset.reset_index(inplace=True)

# Text Representation

This section experiments with different document representations, namely:
- Bag of Words
- N-grams
- TF-IDF
- CBOW
- Skip-gram
- Pre-trained Word2Vec model by Google


Each of the representations are testing using Naive Bayes, and the high performance models are used in the next stage of the pipeline.

Note: Arbitrarily using lemmatization and textblob sentiments

In [36]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.pipeline import Pipeline


In [37]:
X = updated_dataset['Lemmatized Text']
y = updated_dataset['Encoded_TextBlob_Sentiment']
X_train, X_test, y_train, y_test = train_test_split(
    updated_dataset['Lemmatized Text'], 
    updated_dataset['Encoded_TextBlob_Sentiment'],
    test_size=0.2, 
    random_state=42, 
    stratify=updated_dataset['Encoded_TextBlob_Sentiment']
)

## Bag Of Words

In [38]:


bow = CountVectorizer(
    ngram_range= (1,1), # unigram
    preprocessor= lambda x: x, # override preprocessing
    tokenizer= lambda x: x, # override tokenization
)

nb_bow = Pipeline([
    ('bow', bow),
    ('naive_bayes', MultinomialNB())
])



In [39]:
nb_bow.fit(X_train, y_train)
y_pred_bow = nb_bow.predict(X_test)
print(classification_report(y_test, y_pred_bow))

              precision    recall  f1-score   support

          -1       0.76      0.27      0.40        96
           0       0.84      0.60      0.70       455
           1       0.67      0.92      0.78       529

    accuracy                           0.73      1080
   macro avg       0.76      0.60      0.63      1080
weighted avg       0.75      0.73      0.71      1080



## N-Grams
How to find best n-gram number?

In [40]:
ngram = CountVectorizer(
    ngram_range= (1,3), # unigram
    preprocessor= lambda x: x, # override preprocessing
    tokenizer= lambda x: x, # override tokenization
)

ngram_nb = Pipeline([
    ('ngram', ngram),
    ('naive_bayes', MultinomialNB())
])

ngram_nb.fit(X_train, y_train)
ngram_nb_y_pred = ngram_nb.predict(X_test)
print(classification_report(y_test, ngram_nb_y_pred))

              precision    recall  f1-score   support

          -1       0.76      0.32      0.45        96
           0       0.86      0.63      0.73       455
           1       0.69      0.92      0.79       529

    accuracy                           0.75      1080
   macro avg       0.77      0.63      0.66      1080
weighted avg       0.77      0.75      0.73      1080



## TF-IDF

In [42]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(
    preprocessor= lambda x: x, # override preprocessing
    tokenizer= lambda x: x, # override tokenization
)

tfidf_nb = Pipeline([
    ('tfidf', tfidf),
    ('naive_bayes', MultinomialNB())
])

tfidf_nb.fit(X_train, y_train)
tfidf_nb_y_pred = tfidf_nb.predict(X_test)
print(classification_report(y_test, tfidf_nb_y_pred))


              precision    recall  f1-score   support

          -1       0.90      0.09      0.17        96
           0       0.87      0.56      0.68       455
           1       0.65      0.95      0.77       529

    accuracy                           0.71      1080
   macro avg       0.81      0.54      0.54      1080
weighted avg       0.77      0.71      0.68      1080



# Word2Vec

In [49]:
from gensim.models import Word2Vec
import gensim
import numpy as np

### CBOW


Possible parameters to tweak is window, cbow

In [154]:
cbow = Word2Vec(
    sentences=X,
    window=10, # how many context words before and after target word to consider
    min_count=2,
    workers=4, # number of threads
    sg=0, # 0 to use cbow, 1 for skipgram
    cbow_mean=0, # 0 uses sum of context word vectors, 1 uses mean
)
cbow.train(X, total_examples=cbow.corpus_count,epochs=cbow.epochs)
# After training, KeyedVector object is used
cbow_model = cbow.wv

In [118]:
from sklearn.preprocessing import MinMaxScaler

def normalize_vectors(input_vectors):
    """Normalizes vectors to be in range [0,1)
    
    Args:
        input_vectors (np.ndarray): List of vectors, each vector is np.ndarray
    """

    return MinMaxScaler().fit(input_vectors).transform(input_vectors)

In [100]:
def vectorize_sentence(sentence, wv_model):
    """Convert a sentence to a vector by summing word vectors in the sentence.

    Another option could be to average word vectors rather than just sum.
    
    Arg: 
        sentence (list): Tokenized sentence
    Returns:
        (np.ndarray): Vector representation of sentence
    """
    vector_size = wv_model.vector_size
    wv_res = np.zeros(vector_size)
    count = 0
    for w in sentence:
        if w in wv_model:
            # count += 1
            wv_res += wv_model[w]
    # wv_res = wv_res/count
    return wv_res

In [119]:
# Get the embeddings of train and test set
X_train_cbow = normalize_vectors(np.array([vectorize_sentence(x,cbow_model) for x in X_train]))
X_test_cbow = normalize_vectors(np.array([vectorize_sentence(x,cbow_model) for x in X_test]))

#test NB
nb_cbow = MultinomialNB()
nb_cbow.fit(X_train_cbow, y_train)
y_pred_cbow = nb_cbow.predict(X_test_cbow)
print(classification_report(y_test, y_pred_cbow, zero_division=0))

              precision    recall  f1-score   support

          -1       0.00      0.00      0.00        96
           0       0.75      0.26      0.39       455
           1       0.54      0.94      0.69       529

    accuracy                           0.57      1080
   macro avg       0.43      0.40      0.36      1080
weighted avg       0.58      0.57      0.50      1080



## Skip Gram

In [124]:
skipgram = Word2Vec(
    sentences=X,
    window=10,
    min_count=2,
    workers=4,
    sg=1, # 0 to use cbow, 1 for skipgram
)
# Train the model
skipgram.train(X, total_examples=skipgram.corpus_count,epochs=skipgram.epochs)

# After training, KeyedVector object is used
skipgram_model = skipgram.wv

In [121]:
# Get the embeddings of train and test set
X_train_sg = normalize_vectors(np.array([vectorize_sentence(x,skipgram_model) for x in X_train]))
X_test_sg = normalize_vectors(np.array([vectorize_sentence(x,skipgram_model) for x in X_test]))
#test on NB
nb_sg = MultinomialNB()
nb_sg.fit(X_train_sg, y_train)
y_pred_sg = nb_sg.predict(X_test_sg)
print(classification_report(y_test, y_pred_sg, zero_division=0))

              precision    recall  f1-score   support

          -1       0.00      0.00      0.00        96
           0       0.77      0.35      0.48       455
           1       0.56      0.93      0.70       529

    accuracy                           0.60      1080
   macro avg       0.44      0.42      0.39      1080
weighted avg       0.60      0.60      0.54      1080



## Using Pre-trained Word2Vec Model
Using pre-trained word2vec model from Google, which has 3 million words. Trained on 100 billion words from the google news dataset. Unsure if it uses CBOW or SkipGram or both

From https://thinkingneuron.com/how-to-classify-text-using-word2vec/
Google Model can be downloaded from https://thinkingneuron.com/how-to-classify-text-using-word2vec/#:~:text=download%20link%3A%20https%3A%2F%2Fdrive.google.com%2Ffile%2Fd%2F0b7xkcwpi5kdynlnuttlss21pqmm%2Fedit%3Fusp%3Dsharing

In [29]:
# load the word vectors from Google Model (1GB)
# Will take time because 3 mil words. 
# Download model from https://thinkingneuron.com/how-to-classify-text-using-word2vec/#:~:text=download%20link%3A%20https%3A%2F%2Fdrive.google.com%2Ffile%2Fd%2F0b7xkcwpi5kdynlnuttlss21pqmm%2Fedit%3Fusp%3Dsharing
GoogleModel = gensim.models.KeyedVectors.load_word2vec_format('/mnt/d/Andrea/Heriot-Watt/year-4/f20aa/f20aa-coursework-1/GoogleNews-vectors-negative300.bin', binary=True,)

In [145]:
def vectorize_sentence_2(sentences, wv_model):
    """Convert a list of sentences to a vectors.

    By summing word vectors in the sentence.
    Another option could be to average word vectors rather than just sum.
    
    Arg: 
        sentence (pd.Series): Series consisting of list of tokenized texts 
    Returns:
        (np.ndarray): Vector representation of sentences
    """
    vector_size = wv_model.vector_size
    count = 0
    vector_sentences = []

    for i in range(len(sentences)): 
        sentence = np.zeros(vector_size)
        for word in sentences[i]:
            count+=1
            if word in wv_model:
                sentence += wv_model[word]
        # sentence = sentence/count
        vector_sentences.append(sentence)
    vector_sentences_df = pd.DataFrame(vector_sentences)
    
    return vector_sentences_df

Convert text data to W2V vectors

In [146]:
w2v_data = vectorize_sentence_2(X, GoogleModel)
w2v_data.reset_index(inplace=True, drop=True)
w2v_data['sentiment']=y

In [147]:
w2v_data

0         1         2         3         4         5         6    \
0    -1.517578  2.149384  0.244385  1.057068  0.005615 -0.110962 -1.076660   
1    -2.202881  1.526184  1.248779  0.022171 -0.643616 -0.535217  1.304497   
2    -1.591887  2.016541  1.312225  1.879883 -1.195282 -0.182129  2.341568   
3    -0.187256  0.425644  0.694336  0.746582 -0.830971  0.188629  0.263794   
4    -1.289551  0.470154  0.040894  1.148956 -0.361826 -0.663452  1.364746   
...        ...       ...       ...       ...       ...       ...       ...   
5393 -0.779114  1.713898  0.965210  2.606506 -1.261993  0.792786  0.899628   
5394 -0.779114  1.713898  0.965210  2.606506 -1.261993  0.792786  0.899628   
5395 -0.779114  1.713898  0.965210  2.606506 -1.261993  0.792786  0.899628   
5396 -0.170532  0.555901  0.524551  1.156067 -0.659790  0.719757 -0.405373   
5397 -0.910950  1.570068  0.804808  3.015259 -0.817932 -0.055786 -0.946167   

           7         8         9    ...       290       291       292  \
0    -4.466125  2.398560  1.838623  ... -2.183201  2.587189 -1.779480   
1    -1.866699  1.610382  0.834839  ... -2.435616 -0.429688  0.471130   
2    -2.139465  1.937492  1.381104  ... -1.698547  2.054138 -2.196167   
3    -1.604248  0.192200  0.070190  ... -0.465820  0.378479 -1.248840   
4    -1.629761  0.779358  0.507080  ...  0.206055  2.159424 -0.898987   
...        ...       ...       ...  ...       ...       ...       ...   
5393  0.057434  2.080322  1.345978  ... -0.424438 -0.558105 -2.832397   
5394  0.057434  2.080322  1.345978  ... -0.424438 -0.558105 -2.832397   
5395  0.057434  2.080322  1.345978  ... -0.424438 -0.558105 -2.832397   
5396 -0.833130  0.974854  0.811157  ...  0.787354  1.457672  0.223541   
5397 -2.200195  0.376193  1.857975  ... -1.392151  2.878784 -0.290070   

           293       294       295       296       297       298       299  
0     1.421364 -3.537811 -1.348389 -1.448334 -0.404602  0.051758  4.679199  
1     0.044685 -0.072174  0.426514  1.126831 -0.328308  0.306915  1.507935  
2     0.245865 -0.480743  0.292847 -0.172760 -0.098694  1.208496  0.037415  
3     0.507454 -0.283112  0.090332  0.951538 -0.135376 -0.143707  0.971436  
4     1.384079 -0.363281  0.470703 -1.204590 -1.422424 -0.242188 -0.068848  
...        ...       ...       ...       ...       ...       ...       ...  
5393 -0.131470  1.480103  1.432678 -2.717224 -1.542297  0.498291  0.041809  
5394 -0.131470  1.480103  1.432678 -2.717224 -1.542297  0.498291  0.041809  
5395 -0.131470  1.480103  1.432678 -2.717224 -1.542297  0.498291  0.041809  
5396  0.326981  0.550018 -0.178589  0.676849 -0.140881  0.202820 -0.080078  
5397  1.094849 -1.044220 -1.327454 -2.159546 -0.651855  0.339622  1.169434  

[5398 rows x 300 columns]

In [151]:
# Get train and test values
X_w2v = w2v_data.iloc[:,:-1].values
y_w2v = w2v_data.iloc[:,-1].values

X_w2v_scaled = normalize_vectors(X_w2v)

X_train_w2v, X_test_w2v, y_train_w2v, y_test_w2v = train_test_split(
    X_w2v_scaled, y_w2v, test_size=0.2, random_state=42,
    stratify= y)

In [153]:
nb = MultinomialNB()
nb.fit(X_train_w2v, y_train_w2v)
y_pred_w2v = nb.predict(X_test_w2v)
print(classification_report(y_test_w2v, y_pred_w2v, zero_division=0))



              precision    recall  f1-score   support

          -1       0.00      0.00      0.00        96
           0       0.67      0.63      0.65       455
           1       0.64      0.79      0.70       529

    accuracy                           0.65      1080
   macro avg       0.43      0.47      0.45      1080
weighted avg       0.59      0.65      0.62      1080



# Comparison of Text Representation Models
The different models are compared based on their weighted precision, recall, and f1 score.
Values that are closer to one are optimal.

|Model|Precision|Recall|F-1|
|---|---|---|---|
|Bag Of Words (Unigram)|0.75|0.73|0.71|
|Trigrams|0.77|0.75|0.73|
|TF-IDF|0.77|0.71|0.68|
|CBOW (Trained on scraped tweets corpus)|0.58|0.57|0.5|
|Skip-gram (Trained on scraped tweets corpus)|0.6|0.60|0.54|
|Word2Vec model (Trained on Google News Corpus)|0.59|0.65|0.62|


It was expected that the more sophisticated models - those which take into account the context of words - would perform better. However, based on the results the more primitive approaches performed better. This could be due to our small dataset.

The tri-gram model was the best performer in terms of F-1 score, which is the harmonic mean of the precision and recall values.


